Tighten storage-operator permissions against Salt

gdemonet opened the pull request:
**Component**: salt, storage-operator

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: see conversation in #2634

**Summary**: to do

**Acceptance criteria**: to do


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2634 
Fixes: #1528
Fixes: #2084

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

gdemonet committed:
storage-operator: Tighten ClusterRole permissions

We don't want too open permissions for security reasons that were made
obvious in #2634. We thus reduce the rules deployed to the bare minimum,
and add some inline comments to better explain why each ruleset is
needed.

Fixes: #2084

gdemonet committed:
salt, storage-operator: Limit allowed Salt states

Reduce to only state formulas in `metalk8s.volumes`.

Fixes: #1528

gdemonet committed:
tests: Add negative RBAC test against SaltAPI

This checks that if not allowed in Salt eauth perms, an authenticated
user (using 'storage-operator' as a practical example) cannot run
any module (using 'test.ping' as an example).

See: #2634

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.6/bugfix/tighten-storage-operator-perms` with contents from `w/2.5/bugfix/tighten-storage-operator-perms`
and `development/2.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.6/bugfix/tighten-storage-operator-perms origin/development/2.6
 $ git merge origin/w/2.5/bugfix/tighten-storage-operator-perms
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.6/bugfix/tighten-storage-operator-perms
```

slaperche-scality commented on storage-operator/deploy/role.yaml:
```diff
@@ -4,36 +4,23 @@ metadata:
   creationTimestamp: null
   name: storage-operator
 rules:
+# For recording transition events
```
Good idea to add comments in this file :+1:

NicolasT replied:
FWIW, if we update `operator-sdk`/`controller-runtime` (or move to `kubebuilder`), the RBAC rules are to be defined in the Go code using annotations, and this file is to be generated by the tooling.

gdemonet replied:
Oh yeah... Debt ticket #2644

slaperche-scality commented on salt/metalk8s/salt/master/files/master-99-metalk8s.conf.j2:
```diff
@@ -45,7 +45,9 @@ external_auth:
       - '*':
         - 'disk.dump'
-        - 'state.sls'
+        - 'state.sls':
```
Ho, this was simpler than I thought ^^

gdemonet replied:
That's the theory, not working atm :sweat_smile:

gdemonet replied:
OK it looks like it's working actually!

slaperche-scality reviewed (dismissed):
LGTM

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/2.5/bugfix/tighten-storage-operator-perms` with contents from `bugfix/tighten-storage-operator-perms`
and `development/2.5`.

Please resolve the conflict on **the integration branch** (`w/2.5/bugfix/tighten-storage-operator-perms`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/2.5/bugfix/tighten-storage-operator-perms
 $ git pull  # or "git reset --hard origin/w/2.5/bugfix/tighten-storage-operator-perms"
 $ git merge origin/development/2.5
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/bugfix/tighten-storage-operator-perms
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/bugfix/tighten-storage-operator-perms
```

bert-e commented:
# History mismatch



Merge commit #3810e4ebd165c1e2516a16cefc110cd24bd584a0 on the integration branch
`w/2.5/bugfix/tighten-storage-operator-perms` is merging a branch which is neither the current
branch `bugfix/tighten-storage-operator-perms` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/tighten-storage-operator-perms` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/bugfix/tighten-storage-operator-perms` with contents from `bugfix/tighten-storage-operator-perms`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/bugfix/tighten-storage-operator-perms origin/development/2.5
 $ git merge origin/bugfix/tighten-storage-operator-perms
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/bugfix/tighten-storage-operator-perms
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.6/bugfix/tighten-storage-operator-perms` with contents from `w/2.5/bugfix/tighten-storage-operator-perms`
and `development/2.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.6/bugfix/tighten-storage-operator-perms origin/development/2.6
 $ git merge origin/w/2.5/bugfix/tighten-storage-operator-perms
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.6/bugfix/tighten-storage-operator-perms
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

NicolasT commented on storage-operator/deploy/role.yaml:
```diff
@@ -4,36 +4,63 @@ metadata:
-  - statefulsets
   verbs:
-  - '*'
+  - get
```
Why does the storage-operator need to read deployments and replicasets?

gdemonet replied:
It comes from `kubemetrics`: when creating the Service, it reads its Pod's owner controller (a ReplicaSet) and this controller's controller (a Deployment) to determine the right `selector` to use, I think. The information is disseminated in several issues and pull requests in `operator-sdk`, and I found most of them when investigating the `Failed to list *unstructured.Unstructured` error logs. Could do some digging up my history if you want, but I can confirm that there were unauthorized errors if I remove those permissions.

gdemonet committed:
ci: Add debug when failed to provision volumes

When changes lead to issues in storage-operator provisioning of Volumes,
it is useful to provide the user with appropriate details directly in
the logs, instead of requiring manual investigations on the platform.

gdemonet commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye gdemonet.
